### ISSUE TITLE
Transpile everything to lowest browser support for "module" bundle

### ISSUE THREAD

**Issue by treshugart** [FezVrasta/popper.js#219]
### CodePen demo

*This is not a bug with functionality, but a bug with the distribution bundles.*

### Steps to reproduce the problem

1. Install webpack2
2. Bundle latest popper
3. Load in older browser

### What is the expected behavior?

I would expect the `package.json` field for `module` to only use the module feature from ES2015 and everything else should be transpiled out to work with the oldest browser you support.

### What went wrong?

Supported browsers error on unsupported features.

### Notes

This is technically a breaking change because introducing `module` causes Webpack 2 to favour it over `main`. In turn, it consumes it as an ES2015 module, however, it doesn't get transpiled unless explicitly specified to do so. This is because transpiling anything from `node_modules` is generally bad practice, so tooling tends to not do it by default.

Philosophical reference: https://github.com/FezVrasta/popper.js/issues/218

**Comment by FezVrasta** [comment]
Thank you for your report. It was using the "jsnext:main" field that served the same purpose but was replaced by "module" later.

The version shipped in "module" is transpiled to ES6 (which is supported by the major browsers).

I'm not sure if "module" is required to be transpiled downb to ES5... I'll investigate

**Comment by kayneb** [comment]
Thanks for raising on my behalf @treshugart.
To be fair, webpack 2 wasn't complaining about the ES6 features, it was the webpack Uglify plugin - it doesn't expect any ES6 to remain after being processed by webpack. Apparently the `harmony` version of the Uglify plugin supports ES6; however that doesn't really solve the issue if people are still supporting older browsers.

**Comment by treshugart** [comment]
> however that doesn't really solve the issue if people are still supporting older browsers.

I think this is the main point to consider. If a build tool is going to prefer this, the file should be in a format that can be supported by the oldest supported browser.

> I'm not sure if "module" is required to be transpiled down to ES5... I'll investigate

I don't think this is something that tools like webpack "require", but it's something that is generally best practice for a couple of reasons:

1. Both Webpack and Rollup transpilation plugins / loaders, by and large, ignore `node_modules` by default.
2. If you ship your source, you're forcing your consumers to know about which future ES features you're using in your source code, and they then need to configure their environment to meet those requirements. This could be different for every single dependency.

**Comment by FezVrasta** [comment]
My idea about `main` vs `module` was that `module` should work without any transpiler (except for `import` and `export`) and `main` should work without any transpiler at all.

I expanded it to:

`module`: Works on modern browsers that support the current JS specification
`main`: Works on old browsers.

But as you are telling me, probably the second part isn't quite right 🤔

I wonder then how someone could import the "non ES5" version of the code if she doesn't care about older browsers...

**Comment by treshugart** [comment]
> My idea about main vs module was that module should work without any transpiler (except for import and export) and main should work without any transpiler at all.

That seems incompatible with what you expanded it to mean.

> But as you are telling me, probably the second part isn't quite right.

I would expect to be able to just `import popper from 'popper.js';`, or whatever, to work in all supported browsers without having to tweak my build system.

> I wonder then how someone could import the "non ES5" version of the code if she doesn't care about older browsers...

I would make this additive to supplying `main` and `module`. For example, I'd provide an `es` folder, or something (like `lib/` for ES5 in some libs) where you transpile to the latest ES spec and consumers could then opt into it doing something like `import popper from 'popper.js/es';`.

I did a quick search and found https://github.com/infernojs/inferno/issues/928. There's probably more out there, but it might shed some light on some thoughts that others have about it.

**Comment by FezVrasta** [comment]
So, yea.. I should change the targets to be:

- popper.js - ESM + ES6, set as `no idea` entry
- popper.umd.js - UMD + ES5, set as `main` antry
- popper.esm.js - ESM + ES5, set as `module` entry

The number of bundles is getting quite long tho... lol

```
▶ ls -l
total 2096
-rw-r--r--  1 fezvrasta  staff   30146 Apr 17 21:08 popper-utils.es5.js
-rw-r--r--  1 fezvrasta  staff   54759 Apr 17 21:08 popper-utils.es5.js.map
-rw-r--r--  1 fezvrasta  staff    9180 Apr 17 21:08 popper-utils.es5.min.js
-rw-r--r--  1 fezvrasta  staff   51882 Apr 17 21:08 popper-utils.es5.min.js.map
-rw-r--r--  1 fezvrasta  staff   28846 Apr 17 21:08 popper-utils.js
-rw-r--r--  1 fezvrasta  staff   54574 Apr 17 21:08 popper-utils.js.map
-rw-r--r--  1 fezvrasta  staff    8941 Apr 17 21:08 popper-utils.min.js
-rw-r--r--  1 fezvrasta  staff   51747 Apr 17 21:08 popper-utils.min.js.map
-rw-r--r--  1 fezvrasta  staff   71399 Apr 18 16:44 popper.es5.js
-rw-r--r--  1 fezvrasta  staff  114426 Apr 18 16:44 popper.es5.js.map
-rw-r--r--  1 fezvrasta  staff   18461 Apr 18 16:44 popper.es5.min.js
-rw-r--r--  1 fezvrasta  staff  108576 Apr 18 16:44 popper.es5.min.js.map
-rw-r--r--  1 fezvrasta  staff   68897 Apr 18 16:44 popper.js
-rw-r--r--  1 fezvrasta  staff  114225 Apr 18 16:44 popper.js.map
-rw-r--r--  1 fezvrasta  staff   17459 Apr 18 16:44 popper.min.js
-rw-r--r--  1 fezvrasta  staff  108476 Apr 18 16:44 popper.min.js.map
-rw-r--r--  1 fezvrasta  staff   15600 Apr 18 16:44 tooltip.es5.js
-rw-r--r--  1 fezvrasta  staff   19849 Apr 18 16:44 tooltip.es5.js.map
-rw-r--r--  1 fezvrasta  staff    4697 Apr 18 16:44 tooltip.es5.min.js
-rw-r--r--  1 fezvrasta  staff   18959 Apr 18 16:44 tooltip.es5.min.js.map
-rw-r--r--  1 fezvrasta  staff   13464 Apr 18 16:44 tooltip.js
-rw-r--r--  1 fezvrasta  staff   19864 Apr 18 16:44 tooltip.js.map
-rw-r--r--  1 fezvrasta  staff    4038 Apr 18 16:44 tooltip.min.js
-rw-r--r--  1 fezvrasta  staff   18955 Apr 18 16:44 tooltip.min.js.map
```

**Comment by FezVrasta** [comment]
Or maybe, someone will want to use `UMD` with ES6 and in this case it should be:

- popper.js - ESM + ES6
- popper.umd.js - UMD + ES6
- popper.esm.js - ESM + ES6
- popper.es5.js - ESM + ES6
- popper.umd.es5.js - UMD + ES6
- popper.esm.es5.js - ESM + ES6

I mean... I hope you understand my confusion..

**Comment by treshugart** [comment]
> The number of bundles is getting quite long tho... lol

I know what you mean. The amount of tooling you have to maintain becomes difficult as a result.

> Or maybe, someone will want to use UMD with ES6 and in this case it should be:

I'm not sure you need to cater for that. Someone may want to target new features, but you have to consider whether or not that's actually worth your time. What benefit are they *really* getting from it. If it outweighs your maintenance cost, then maybe it is.

The way I look at it is:

- `main` - UMD
- `module` - ES2015 modules + ES5 everything else

This is also now [nwb](https://github.com/insin/nwb) interprets it. It outputs an `es` folder that contains modules + ES5, `lib` that contains Common JS and a `umd` folder that has UMD. I tend to disagree with the `lib` folder even being necessary, but that's tangental to what we're discussing here.

**Comment by jameslnewell** [comment]
There's a really interesting convo on the topic here https://twitter.com/Rich_Harris/status/852264562242072576

tl;dr

It'd be really cool to let library users transpile libraries to whatever target they want (think babel-preset-env), but the tooling isn't there yet, keep shipping ES5 for `.main` and ES5+ESM  for `.module`. Interestingly, Angular and other folks are pushing `.es2015` for untranspiled ES2015.

**Comment by treshugart** [comment]
@jameslnewel Thanks for linking that!

> Angular and other folks are pushing .es2015 for untranspiled ES2015.

That would be unfortunate. l I think a better solution would be `esLatest`. As specs are finalised and shipped, the key doesn't change but the targets can. Either way it's a breaking API change when you decide to target a later version, probably best not to change (or add) the `package.json` key every year.

**Comment by jameslnewell** [comment]
Hmm I still haven't mastered twitter ☺️ Here's some more of the replies I was referring to https://twitter.com/Rich_Harris/status/852264367911469059

**Comment by FezVrasta** [comment]
Is it okay if I change `main` to `dist/popper.umd.js` os is it considered a breaking change? 🤔

**Comment by jameslnewell** [comment]
Is `dist/popper.umd.js` ES5 if its referred to by `.main`?

Yep, that's great reasoning provided the compile target hasn't changed!

**Comment by FezVrasta** [comment]
Please review the linked PR for more info and give me your feedbacks.

**Comment by jameslnewell** [comment]
> In `dist/umd` there's the UMD + ES5 code.

That means `.main` should be `dist/umd/popper.js`, not `dist/popper.umd.js`?

**Comment by jameslnewell** [comment]
ah nvm, its correct in the PR :)

**Comment by FezVrasta** [comment]
Is it good to merge then?

**Comment by jameslnewell** [comment]
Looks like it should work to me. 👍

However, I didn't run into the issue, I just saw the convo at work and thought the Twitter thread I had just read was relevant to the discussion. The guys will be back in the office tomorrow if you'd like them to test it via a `rc` release.

Thanks for your quick turn around!

**Comment by FezVrasta** [comment]
Okay cool, I'll wait for him 👍

**Comment by FezVrasta** [comment]
released as 1.8.2
